flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/exceptionhistory/FailureHandlingResultSnapshot.java

```diff
@@ -0,0 +1,146 @@
+import org.apache.flink.util.Preconditions;
+
+import javax.annotation.Nullable;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+/**
+ * {@code FailureHandlingResultSnapshot} creates a snapshot of a {@link FailureHandlingResult}
+ * providing the actual {@link Execution Executions}.
+ */
+public class FailureHandlingResultSnapshot {
+
+    private final ExecutionAttemptID rootCauseExecution;
+    private final Throwable rootCause;
+    private final long timestamp;
+    private final Map<ExecutionAttemptID, Execution> failedExecutions;
+
+    /**
+     * Creates a {@code FailureHandlingResultSnapshot} based on the passed {@link
+     * FailureHandlingResult} and {@link ExecutionVertex ExecutionVertices}.
+     *
+     * @param failureHandlingResult The {@code FailureHandlingResult} that is used for extracting
+     *     the failure information.
+     * @param executionVertices The {@code ExecutionVertices} that shall be scanned for failures
+     *     based on the passed {@code FailureHandlingResult}.
+     * @return The {@code FailureHandlingResultSnapshot}.
+     */
+    public static FailureHandlingResultSnapshot create(
+            FailureHandlingResult failureHandlingResult,
+            Iterable<ExecutionVertex> executionVertices) {
+        final ExecutionVertexID rootCauseExecutionVertexId =
+                failureHandlingResult.getExecutionVertexIdOfFailedTask().orElse(null);
+
+        ExecutionAttemptID rootCauseExecutionAttemptId = null;
+        final Map<ExecutionAttemptID, Execution> failedExecutions = new HashMap<>();
+        for (ExecutionVertex executionVertex : executionVertices) {
+            if (executionVertex.getID().equals(rootCauseExecutionVertexId)) {
+                Preconditions.checkArgument(
+                        executionVertex.getCurrentExecutionAttempt().getFailureInfo().isPresent(),
+                        String.format(
+                                "The execution %s didn't provide a failure info even though the corresponding ExecutionVertex %s is marked as having handled the root cause of this failure.",
+                                executionVertex.getCurrentExecutionAttempt().getAttemptId(),
+                                executionVertex.getID()));
+                rootCauseExecutionAttemptId =
+                        executionVertex.getCurrentExecutionAttempt().getAttemptId();
+            }
+
+            final Execution execution = executionVertex.getCurrentExecutionAttempt();
+            if (failureHandlingResult.getVerticesToRestart().contains(executionVertex.getID())
+                    && execution.getFailureInfo().isPresent()) {
+                failedExecutions.put(execution.getAttemptId(), execution);
+            }
+        }
+
+        return new FailureHandlingResultSnapshot(
+                rootCauseExecutionAttemptId,
+                failureHandlingResult.getError(),
```

Comment:
`getError` returns a field which is `@Nullable`. This violates the constructor.

Reply:
I added a workaround setting the exception explicitly to have a working version. The issue in general is again what's already covered by [FLINK-22060](https://issues.apache.org/jira/browse/FLINK-22060). Without the workaround, `JobMasterTest` is failing. AFAIR, it is a known issue that came up while doing the research on `FLINK-22060`. The `JobMasterTest` does not set the failure cause (because it didn't need to before).

The workaround fixes it for now and we should address the issue in a follow up ticket before releasing.

Reply:
> I added a workaround setting the exception explicitly to have a working version. The issue in general is again what's already covered by FLINK-22060. Without the workaround, JobMasterTest is failing. AFAIR, it is a known issue that came up while doing the research on FLINK-22060. The JobMasterTest does not set the failure cause (because it didn't need to before).
> The workaround fixes it for now and we should address the issue in a follow up ticket before releasing.

I re-iterated over the `@Nullable` constraints and updated the code accordingly. We're relying on a utility method `ErrorInfo.handleMissingThrowable` that addresses the issue of [FLINK-21376](https://issues.apache.org/jira/browse/FLINK-21376)/[FLINK-22060](https://issues.apache.org/jira/browse/FLINK-22060) for now.